bootenv/keydata.go

```diff
@@ -0,0 +1,538 @@
+
+	if !str.ReadASN1ObjectIdentifier(&oid) {
+		return 0, errors.New("malformed Algorithm identifier")
+	}
+
+	var null uint8
+
+	if !str.ReadUint8(&null) {
+		return 0, errors.New("malformed input")
+	}
+
+	if len(oid) == len(sha1Oid) {
+		return hashAlg(crypto.SHA1), nil
+	}
+
+	switch oid[8] {
+	case sha224Oid[8]:
+		return hashAlg(crypto.SHA224), nil
+	case sha256Oid[8]:
+		return hashAlg(crypto.SHA256), nil
+	case sha384Oid[8]:
+		return hashAlg(crypto.SHA384), nil
+	case sha512Oid[8]:
+		return hashAlg(crypto.SHA512), nil
+	default:
+		return 0, errors.New("unsupported hash algorithm")
+	}
+}
+
+// digestList corresponds to a list of digests.
+type digestList struct {
+	Alg     hashAlg  `json:"alg"`     // The digest algorithm
+	Digests [][]byte `json:"digests"` // The list of digests
+}
+
+func (l *digestList) marshalASN1WithTag(tag cryptobyte_asn1.Tag, b *cryptobyte.Builder) {
+	b.AddASN1(tag, func(b *cryptobyte.Builder) { // DigestList ::= SEQUENCE {
+		l.Alg.marshalASN1(b)                                         // algorithm AlgorithmIdentifier
+		b.AddASN1(cryptobyte_asn1.SET, func(b *cryptobyte.Builder) { // digests Digests
+			for _, digest := range l.Digests {
+				b.AddASN1OctetString(digest)
+			}
+		})
+	})
+}
+
+type scopeParams struct {
+	ModelDigests digestList `json:"model_digests,omitempty"`
+	Modes        []string   `json:"modes,omitempty"`
+}
+
+func (s *scopeParams) marshalASN1(b *cryptobyte.Builder) {
+	b.AddASN1(cryptobyte_asn1.SEQUENCE, func(b *cryptobyte.Builder) { // Scope ::= SEQUENCE {
+		if len(s.ModelDigests.Digests) > 0 {
+			s.ModelDigests.marshalASN1WithTag(cryptobyte_asn1.Tag(0).ContextSpecific().Constructed(), b) // modelDigests [0] IMPLICIT DigestList OPTIONAL
+		}
+		if len(s.Modes) > 0 {
+			b.AddASN1(cryptobyte_asn1.Tag(1).ContextSpecific().Constructed(), func(b *cryptobyte.Builder) { // modes [1] IMPLICIT BootModes OPTIONAL
+				for _, mode := range s.Modes {
+					b.AddASN1(cryptobyte_asn1.UTF8String, func(b *cryptobyte.Builder) {
+						b.AddBytes([]byte(mode))
+					})
+				}
+			})
+		}
+	})
+}
+
+type ecdsaPublicKey struct {
+	*ecdsa.PublicKey
+}
+
+func (k ecdsaPublicKey) MarshalJSON() ([]byte, error) {
+	der, err := x509.MarshalPKIXPublicKey(k.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+	return json.Marshal(der)
+}
+
+func (k *ecdsaPublicKey) UnmarshalJSON(data []byte) error {
+	var der []byte
+	if err := json.Unmarshal(data, &der); err != nil {
+		return err
+	}
+	pubKey, err := x509.ParsePKIXPublicKey(der)
+	if err != nil {
+		return err
+	}
+	ecdsaKey, isECDSA := pubKey.(*ecdsa.PublicKey)
+	if !isECDSA {
+		return errors.New("invalid key type")
+	}
+	k.PublicKey = ecdsaKey
+	return nil
+}
+
+type keyDataScope struct {
+	Version int `json:"version"`
+
+	Params    scopeParams    `json:"params"`
+	Signature []byte         `json:"signature"`
+	PublicKey ecdsaPublicKey `json:"pubkey"`
+
+	KDFAlg hashAlg `json:"kdf_alg"`
+	MDAlg  hashAlg `json:"md_alg"`
+}
+
+type AdditionalData struct {
```

Comment:
Probably not, I made it private again. I believe why I made it public in the first place was because I thought that platform implementations would need to refer to it but I see now that the TPM one brings its own, the upcoming plainkey too. @chrisccoulson am I missing something? Does the spec reflect that?

Reply:
There shouldn't be a need to access this externally - the only thing that is needed externally is the serialized form of it (via `MakeAdditionalData`), and the format of this is a private implementation detail.